lib/crawler.js

```diff
@@ -324,32 +388,134 @@ Crawler.prototype._buildHttpRequest = function _buildHTTPRequest (options) {
+				self.http2Connections[targetHost] = http2.connect(targetHost);
 
-		var requestArgs = ['uri','url','qs','method','headers','body','form','formData','json','multipart','followRedirect','followAllRedirects','maxRedirects','removeRefererHeader','encoding','pool','timeout','proxy','auth','oauth','strictSSL','jar','aws','gzip','time','tunnel','proxyHeaderWhiteList','proxyHeaderExclusiveList','localAddress','forever', 'agent', 'strictSSL', 'agentOptions', 'agentClass'];
+				log('debug',`connect to a new ${targetHost}`);
+				self.http2Connections[targetHost].on('error', (err) => {
+					http2Error = err;
+
+					if(self.http2Connections[targetHost]) self._closeAndDeleteHttp2Session(targetHost);
+				});
+
+				self.http2Connections[targetHost].on('frameError', (err) => {
+					http2Error = err;
+				});
 
-		request(_.pick.apply(self,[ropts].concat(requestArgs)), function(error,response) {
-			if (error) {
-				return self._onContent(error, options);
+				self.http2Connections[targetHost].on('goaway', () => {
+					log('debug',`${targetHost} connection goaway`);
+				});
+
+				self.http2Connections[targetHost].on('connect', () => {
+					log('debug',`${targetHost} connection init`);
+				});
+
+				self.http2Connections[targetHost].on('timeout', (err) => {
+					http2Error = err;
+				});
+
+				self.http2Connections[targetHost].on('close', () => {
+					log('debug',`${targetHost} connection closed`);
+					if(self.http2Connections[targetHost])  self._closeAndDeleteHttp2Session(targetHost);
+				});
 			}
 
-			self._onContent(error,options,response);
-		});
+			const currentHttp2Session = self.http2Connections[targetHost];
+			let req;
+
+			try {
+				req = currentHttp2Session.request(ropts.headers);
+			} catch (e) {
+				//to handle the goaway issue， goaway will make the session can not be established
+				//but it can not be detected at the moment that stream init
+				//try catch seems the way to sovle it
+				self._onContent(e, options, response);
+				return;
+			}
+
+			req.on("response", (headers, flags) => {
+				//Where build the response obj
+				response['statusCode'] = headers[':status'];
+				response['request'] = {};
+				response['request']['uri'] = options.uri;
+				response['request']['method'] = options.method;
+				response['request']['headers'] = options.headers;
+				for (const name in headers) {
+					response['headers'][name] = headers[name];
+				}
+			});
+
+			req.on('error', (err) => {	
+				log('error', `inside http2 stream, got error`);
+				http2Error = err;
+			})
+
+			req.on("data", chunk => {
+				chunks.push(chunk);
+			});
+
+			req.on("end", () => {
+				if (http2Error) self._onContent(http2Error, options, response);
+				else {
+					response.body = Buffer.concat(chunks);
+					self._onContent(null, options, response);
+				}
+			});
+
+			//set request body
+			req.end(requestBody);
+
+			if (self.options.timeout !== undefined && self.options.timeout > 0) {
```

Comment:
why place this code block here? what's the difference between other events? if no, better to put them together.

Reply:
first judgement may be redundant